[ESS] BYO LLM guide (#5530)
elastic/security-docs: docs/AI-for-security/connect-to-byo.asciidoc

[[connect-to-byo-llm]]
= Connect to your own local LLM

:frontmatter-description: Set up a connector to LM Studio so you can use a local model with AI Assistant.
:frontmatter-tags-products: [security]
:frontmatter-tags-content-type: [guide]
:frontmatter-tags-user-goals: [get-started]

> This page provides instructions for setting up a connector to a large language model (LLM) of your choice using LM Studio. This allows you to use your chosen model within {elastic-sec}. You'll first need to set up a reverse proxy to communicate with {elastic-sec}, then set up LM Studio on a server, and finally configure the connector in your {elastic-sec} project. https://www.elastic.co/blog/ai-assistant-locally-hosted-models[Learn more about the benefits of using a local LLM].

Reviewer: ```suggestion
This page provides instructions for setting up a connector to a large language model (LLM) of your choice using LM Studio. This allows you to use your chosen model within {elastic-sec}. You'll first need to set up a reverse proxy to communicate with {elastic-sec}, then set up LM Studio on a server, and finally configure the connector in {kib}. https://www.elastic.co/blog/ai-assistant-locally-hosted-models[Learn more about the benefits of using a local LLM].
```
As Joe suggested yesterday, we could probably also replace this with a less platform-specific term like "Elastic deployment/instance".

The same suggestion applies to lines 176, 178, and 180.

    add_header x-xss-protection "1; mode=block" always;
    add_header x-frame-options "SAMEORIGIN" always;
    add_header X-Content-Type-Options "nosniff" always;
    add_header Referrer-Policy "strict-origin-when-cross-origin" always;
    ssl_stapling on;
    ssl_stapling_verify on;
    ssl_trusted_certificate /etc/letsencrypt/live/<yourdomainname.com>/fullchain.pem;
    resolver 1.1.1.1;
    location / {
          
    if ($http_authorization != "Bearer <secret token>") {
    return 401;
}                            

    proxy_pass http://localhost:1234/;
   }

}
--------------------------------------------------

IMPORTANT: If using the example configuration file above, you must replace several values: Replace `<secret token>` with your actual token, and keep it safe since you'll need it to set up the {elastic-sec} connector. Replace `<yourdomainname.com>` with your actual domain name. Update the `proxy_pass` value at the bottom of the configuration if you decide to change the port number in LM Studio to something other than 1234.

[discrete]
=== (Optional) Set up performance monitoring for your reverse proxy

> You can use Elastic's https://www.elastic.co/docs/current/integrations/nginx[Nginx integration] to monitor performance and populate monitoring dashboards in the {security-app}.

Reviewer: ```suggestion
You can use Elastic's {integrations-docs}/nginx[Nginx integration] to monitor performance and populate monitoring dashboards in the {security-app}.
```

After the model loads, you should see a `Model loaded successfully` message in the CLI.

image::images/lms-studio-model-loaded-msg.png[The CLI message that appears after a model loads]

To verify which model is loaded, use the `lms ps` command.

image::images/lms-ps-command.png[The CLI message that appears after running lms ps]

If your model uses NVIDIA drivers, you can check the GPU performance with the `sudo nvidia-smi` command.

[discrete]
=== Option 2: load a model using the GUI

Refer to the following video to see how to load a model using LM Studio's GUI. You can change the **port** setting, which is referenced in the Nginx configuration file. Note that the **GPU offload** was set to **Max**.

=======
++++
<script type="text/javascript" async src="https://play.vidyard.com/embed/v4.js"></script>
<img
  style="width: 100%; margin: auto; display: block;"
  class="vidyard-player-embed"
  src="https://play.vidyard.com/FMx2wxGQhquWPVhGQgjkyM.jpg"
  data-uuid="FMx2wxGQhquWPVhGQgjkyM"
  data-v="4"
  data-type="inline"
/>
</br>
++++
=======

[discrete]
== (Optional) Collect logs using Elastic's Custom Logs integration

> You can monitor the performance of the host running LM Studio using Elastic's https://www.elastic.co/docs/current/integrations/log[Custom Logs integration]. This can also help with troubleshooting. Note that the default path for LM Studio logs is `/tmp/lmstudio-server-log.txt`, as in the following screenshot:

Reviewer: ```suggestion
You can monitor the performance of the host running LM Studio using Elastic's {integrations-docs}/log[Custom Logs integration]. This can also help with troubleshooting. Note that the default path for LM Studio logs is `/tmp/lmstudio-server-log.txt`, as in the following screenshot:
```

image::images/lms-custom-logs-config.png[The configuration window for the custom logs integration]

[discrete]

> == Configure the connector in {elastic-sec}
> 
> Finally, configure the connector in your Security project:
> 
> 1. Log in to your Security project.

Author: ```suggestion
== Configure the connector in your Elastic deployment

Finally, configure the connector:

1. Log in to your Elastic deployment.
```
